06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/TicketDepot/ticketdepot.spec

```diff
@@ -0,0 +1,27 @@
+
+methods {
+    transactionFee() returns (uint64) envfree
+    owner() returns (address) envfree
+}
+
+ghost uint256 numberOfEvents;
+
+hook Sstore numEvents uint16 v STORAGE {
+    numberOfEvents = v;
+}
+
+
+// only createEvent can change the number of events
+rule createEventIsOnlyFunctionThatCanIncreaseNumEvents() {
```

Comment:
I do not understand how ghosts work, because this rule is finding a counterexample when I think it shouldn't. See this Verification Report: https://prover.certora.com/output/20739/e0ec12bc314c434b95dd66284cfb9343?anonymousKey=b54c4f5c83b5586bce3dbc6ac912b5c21544a159

where `nEventsBefore` starts at 15, then `f(e, args)` gets called (where `f` is `createEvent`), **but then `nEventsAfter` is still 5**. Looking at the code for `TicketDepot.createEvent`, this seems impossible. How would `numberOfEvents` not be updated via the `createEvent` call?

Reply:
The tool can choose any value for nEvents ghost before the actual variable is touched and the hook is triggered which updates the ghost. In this case the tool chose to make nEvents 15 while the actual variable was 14, the ghost was "updated" to the real value 15 after the call but didn't change since the tool had chosen 15 as the value to start with. You don't need a ghost in this situation as you're not tracking anything that you can't access directly in the code. If you want to use a ghost, you can require before anything that it is equal to the actual variable it is tracking.

Reply:
> You don't need a ghost in this situation as you're not tracking anything that you can't access directly in the code.

But *I am* tracking something I can't directly access in the code, specifically the `TicketDepot.numEvents` which is defined on line 19 as

`uint16 numEvents;`

and so it will not have a getter automatically generated for it (a getter would only be generated if it was explicitly marked `uint16 public numEvents;`).

So same question, how would I use a ghost to keep track of `numEvents`, and have both the ghost variable `numberOfEvents` and the actual storage variable `numEvents` equal each other all the time?

Reply:
Ok that's my mistake I thought I checked the code and saw that the variable was declared public. I think I checked the original repo instead of yours. In this case the reason the value of the ghost and the variable are not matching is because the hook that requires they are the same is triggered on Sstores which is specified at the declaration of the hook. I believe you can fix this by adding another hook that is triggered on Sloads.